app/controllers/spree/orders_controller_decorator.rb

```diff
@@ -202,4 +223,23 @@ def populate_cart hash
   def wrap_json_infinity(n)
     n == Float::INFINITY ? 2147483647 : n
   end
+
+  def order_to_update
+    order = Spree::Order.complete.find_by_number(params[:id])
+    return order if order.andand.changes_allowed? && can?(:update, order)
+    current_order
+  end
```

Comment:
It's not really authorisation logic, so I don't think it belongs with CanCan. It's trying to work out which order the user wants to update. `current_order` is defined on the session (it's basically the 'cart', which is always present), but if the user submits a specific order `id` with their update request, this method helps interpret that they want to update that order instead of the default.

You're right though, should probably memoize the result.